syntax = "proto3";

import "google/protobuf/any.proto";

message Message {
    NodeInfoProto senderInfo = 1;
    google.protobuf.Any message = 2;
}

message NodeInfoProto {
    enum NodeHealth {
        DEAD = 0;
        SUSPECT = 2;
        ALIVE = 4;
    }
    string name = 1;
    string host = 2;
    int32 port = 3;
    int32 incarnation = 4;
    NodeHealth health = 5;
    string gossipHost = 6;
    int32 gossipPort = 7;
}

message Ping {
    NodeInfoProto targetInfo = 1;
}

message PingReq {
    NodeInfoProto targetInfo = 1;
}

message Ack {}

message JoinRequest {}

message JoinAccept {
    NetworkView networkView = 1;
}

message Gossip {
    NodeInfoProto originNode = 1;
    int32 remainingSends = 2;
    string gossipId = 3;
    google.protobuf.Any message = 4;
}

message NetworkView {
    repeated NodeInfoProto nodes = 1;
}



message ServiceRequest {
    int32 request_id = 1;
    string serviceName = 3;
    google.protobuf.Any message = 4;
}

message ServiceResponse {
    int32 request_id = 1;
    string serviceName = 3;
    google.protobuf.Any message = 4;
}

message AddRequest {
    int32 numA = 1;
    int32 numB = 2;
}


message AddResponse {
    int32 result = 1;
}

message SimpleString {
    string aString = 1;
}
